### PR TITLE
Correctly emit imported 'React' references in JSX.

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes bug #4422

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -869,11 +869,13 @@ namespace ts {
         _openingElementBrand?: any;
         tagName: EntityName;
         attributes: NodeArray<JsxAttribute | JsxSpreadAttribute>;
+        reactNode?: Identifier;
     }
 
     /// A JSX expression of the form <TagName attrs />
     export interface JsxSelfClosingElement extends PrimaryExpression, JsxOpeningElement {
         _selfClosingElementBrand?: any;
+        reactNode?: Identifier;
```
Comment (sheetalkamat) [inline src/compiler/types.ts]: this isn't needed right? its getting it from JsxOpeningElement right ?

File: tests/baselines/reference/tsxReactEmit5.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+//// [react-consumer.tsx]
+import {React} from "./test";
+// Should emit test_1.React.createElement
+var spread1 = <div />;
+
+//// [file.js]
+//// [react-consumer.js]
+var test_1 = require("./test");
+// Should emit test_1.React.createElement
+var spread1 = test_1.React.createElement("div", null);
```
Comment (sheetalkamat) [inline tests/baselines/reference/tsxReactEmit5.js]: can you add testcase for React.spread too

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -7629,8 +7629,12 @@ namespace ts {
             // is no 'React' symbol in scope, we should issue an error.
             if (compilerOptions.jsx === JsxEmit.React) {
                 let reactSym = resolveName(node.tagName, "React", SymbolFlags.Value, Diagnostics.Cannot_find_name_0, "React");
-                if (reactSym) {
+                if (reactSym && reactSym !== unknownSymbol) {
                     getSymbolLinks(reactSym).referenced = true;
+                    let reactNode = <Identifier>createSynthesizedNode(SyntaxKind.Identifier, false);
+                    reactNode.text = 'React';
+                    getNodeLinks(reactNode).resolvedSymbol = reactSym;
+                    node.reactNode = reactNode;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: i do not think this is going to work with incremental parsing. now we are adding artifacts on the node that will not be removed if the user removed react.

I think the right way to handle this is for the emitter to ask the checker or the emit resolver to be more accurate about it, and the checker could possibly cache it if needed in the nodeLinks, though might not be necessary. 

check with @ahejlsberg as well.

Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Discussed; solution is to emit a synthetic `React` identifier node instead here

**Comment by mhegazy** [comment]
:+1:

**Comment by RyanCavanaugh** [comment]
n.b. Need to cherry-pick this into release-1.6

**Pull request merged**